// @stanvanheumen/ngx-storage/esm5/stanvanheumen-ngx-storage.js
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import { _throw } from 'rxjs/observable/throw';
import { Injectable, NgModule } from '@angular/core';

var StorageService = /** @class */ (function () {
    function StorageService() {
        this.storage = (typeof window !== 'undefined' && window.hasOwnProperty('localStorage'))
            ? window.localStorage
            : null;
        this.state = {};
        this.warn = function (text) {
            if (console && console.warn) {
                console.warn(text);
            }
        };
        if (!this.isStorageAvailable) {
            this.warn('The local storage is not available on this platform.');
        }
    }
    StorageService.prototype.get = function (token) {
        if (this.state[token]) {
            return this.state[token];
        }
        var rawData = null;
        if (this.isStorageAvailable) {
            rawData = this.storage.getItem(token);
        }
        var parsedData = null;
        if (rawData !== null) {
            try {
                parsedData = JSON.parse(rawData);
            }
            catch (error) {
                return _throw(new Error('The raw data cannot be parsed; Invalid data in the storage.'));
            }
        }
        this.state[token] = new BehaviorSubject(parsedData);
        return this.state[token];
    };
    StorageService.prototype.set = function (token, data) {
        var rawData = JSON.stringify(data);
        if (this.isStorageAvailable) {
            this.storage.setItem(token, rawData);
        }
        if (!this.state[token]) {
            return;
        }
        var current = JSON.stringify(this.state[token].getValue());
        if (current === rawData) {
            return;
        }
        this.state[token].next(data);
    };
    StorageService.prototype.remove = function (token) {
        if (this.isStorageAvailable) {
            this.storage.removeItem(token);
        }
        if (!this.state[token]) {
            return;
        }
        var current = this.state[token].getValue();
        var next = null;
        if (current === next) {
            return;
        }
        this.state[token].next(next);
    };
    StorageService.prototype.clear = function () {
        var _this = this;
        if (this.isStorageAvailable) {
            this.storage.clear();
        }
        Object.keys(this.state).forEach(function (key) {
            var current = _this.state[key].getValue();
            var next = null;
            if (current === next) {
                return;
            }
            _this.state[key].next(next);
        });
    };
    Object.defineProperty(StorageService.prototype, "isStorageAvailable", {
        get: function () {
            return !!this.storage;
        },
        enumerable: true,
        configurable: true
    });
    return StorageService;
}());
StorageService.decorators = [
    { type: Injectable },
];
StorageService.ctorParameters = function () { return []; };
var NgxStorageModule = /** @class */ (function () {
    function NgxStorageModule() {
    }
    return NgxStorageModule;
}());
NgxStorageModule.decorators = [
    { type: NgModule, args: [{
                providers: [StorageService]
            },] },
];
NgxStorageModule.ctorParameters = function () { return []; };

export { NgxStorageModule, StorageService };
